framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java

```diff
@@ -1198,6 +1205,35 @@ private ExpressionErrorCollector() {
     }
   }
 
+  /**
+   * Replaces a dependent type annotation with a parser error with the top qualifier in the
+   * hierarchy.
+   */
+  protected class ErrorAnnoReplacer extends SimpleAnnotatedTypeScanner<Void, Void> {
+
+    /**
+     * Create ErrorAnnoReplacer
+     *
+     * @param qh qualifier hierarchy
+     */
+    private ErrorAnnoReplacer(QualifierHierarchy qh) {
+      super(
+          (AnnotatedTypeMirror type, Void aVoid) -> {
+            AnnotationMirrorSet newAnnos = new AnnotationMirrorSet();
```

Comment:
This does a lot of allocation.  I suggest first determining whether creating a new AnnotationMirrorSet is necessary, and then doing the work only if so.  This requires two passes over the set, but I think it's worth it given how often this will be called.